src/composables/viewport.ts

```diff
@@ -0,0 +1,51 @@
+import { debounce } from 'lodash-es';
+import { ref } from '@vue/composition-api';
+import { MOBILE_WIDTH } from '../popup/utils';
+
+export interface IScrollCallbackParams {
+  isOutsideOfViewport: boolean
+}
+
+const viewportElement = ref();
+
+export const useViewport = () => {
+  function onViewportScroll(callback: Function) {
+    const isDesktop = (
+      document.documentElement.clientWidth > MOBILE_WIDTH
+      || process.env.IS_EXTENSION
+    );
+
+    const { scrollHeight, scrollTop, clientHeight } = isDesktop
+      ? viewportElement.value
+      : document.documentElement;
+
+    const isOutsideOfViewport = scrollHeight - scrollTop <= clientHeight + 100;
+
+    callback({ isOutsideOfViewport });
+  }
+
+  function initViewport(scrollableElement: HTMLDivElement | null) {
+    viewportElement.value = scrollableElement;
+  }
+
+  function debouncedOnScrollAction(onScrollMethod: Function) {
+    return debounce(() => onViewportScroll(onScrollMethod), 100);
+  }
```

Comment:
In my opinion there is no need to use additional function and pass the arguments to it when attaching the event listeners. You can just create debounced function:

```
const onViewportScroll = debounce(() => {
  // Body of the function
}, 100);

<Elem>.addEventListener('scroll', () => onViewportScroll(callback));
```

Also the `addViewportListener` and `clearViewportListener` can be run inside of this composable to avoid having to run the `onMounted`and `onUnmounted` in a place we want to use it:

So here it can look this way:
```
function onViewportScroll(callback: Function) {
  onMounted(() => addViewportListener(callback));
  onBeforeUnmount(() => clearViewportListener(callback));
}
```

so then in the component running it will look this way:
```
const { onViewportScroll } = useViewport();

onViewportScroll(() => {
  // Local component's code
});
```

Not tested, just an idea.